src/main/java/bio/terra/service/load/LoadDao.java

```diff
@@ -0,0 +1,128 @@
+package bio.terra.service.load;
+
+
+import bio.terra.common.DaoKeyHolder;
+import bio.terra.service.load.exception.LoadLockFailureException;
+import bio.terra.service.load.exception.LoadLockedException;
+import org.apache.commons.codec.binary.StringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.jdbc.core.namedparam.MapSqlParameterSource;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Propagation;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.UUID;
+
+@Repository
+public class LoadDao {
+    private final Logger logger = LoggerFactory.getLogger("bio.terra.service.snapshot.SnapshotDao");
+
+    private final NamedParameterJdbcTemplate jdbcTemplate;
+
+    @Autowired
+    public LoadDao(NamedParameterJdbcTemplate jdbcTemplate) {
+        this.jdbcTemplate = jdbcTemplate;
+    }
+
+    @Transactional(propagation = Propagation.REQUIRED)
+    public Load lockLoad(String loadTag, String flightId) {
+        Load load = lookupLoadByTag(loadTag);
+        if (load == null) {
+            load = createLoad(loadTag, flightId);
+        }
+
+        if (load.isLocked()) {
```

Comment:
If two flights try to lock the load concurrently, is there a chance that both could hit this point with `load.isLocked() == false`? If so, what happens?

Reply:
Yes, both could see `load.isLocked() == false`.  They would then go forward to attempt the update.
We rely on the database concurrency control to ensure that only one will succeed in performing the update. The other transaction will fail. If you want a deeper dive, I can explain how Postgres accomplishes that.

You do point out a problem: the fix needed here is catching the SQL update failure and converting the error into the `LoadLockFailureException`.

Reply:
Writing the code and testing for this revealed that we had not properly hooked in a transaction manager in Spring, so the code was not actually working. So you get bonus points for this comment!